Small refactoring

ShunSakurai opened the pull request:
1. Run a lint and move query parameters to AsanaController.js

I thought editing the code would become easier if we define the query parameters in AsanaController.js. This way, we no longer need to check frequently what parameters are set in AsanaGatewayService.js.

2. Make the project not a required field

I thought it's easier to create tasks if projects are not necessarily the required field.

3. Open the page by clicking the successful task creation status

I found that when creating a task for my Personal Projects workspace, I cannot select me as the user, so when I don't want to add that task without a project, it becomes harder to find the created task later. I added a feature to open the created task right after creating, just as the [official extension](https://chrome.google.com/webstore/detail/asana-extension-for-chrom/khnpeclbnipcdacdkhejifenadikeghk?hl=en-US) does. The next step would be to default 'me' as the assignee if no project or tag is added.

4. Default me as the assignee when creating a task without assignee, project, or tag

[Asana Tip: If you forget to give the task an assignee or project, it will default as assigned to you so you can find it easily in your My Tasks list.](https://asana.com/guide/help/tasks/actions)

5. Add link to the Chrome Web Store page

Comment

I added three more commits to this request later on. Please let me know if I should rebase and merge them all to one commit. If you have any feedback or comments, please let me know. I'm happy to listen to it and edit my contribution. Thanks!

amitg87 commented:
**Run a lint and move query parameters to AsanaController.js**
Running lint is good idea.
I am against moving query params to controller:
1. Separation of concern: controller requests data from service, service is responsible for invoking api. Since these parameters control the response to request  made by service and are not related to controller in any way. Lets keep it as it was.
2. We will refactor service into in future - lets keep this for future.

**Make the project not a required field**
Not specifying project make tracking the task difficult. Well know problem in official extension.
Atleast for I will keep project compulsory.

**Default me as the assignee when creating a task without assignee, project, or tag**
No. Bad Idea. Two reason:
1. Tasks are always part of project. Even if assignee is not fixed, somebody from team takes care of it.
2. We are solving a problem that nobody seems to facing.

**Open the page by clicking the successful task creation status**
Good idea. I missed this one. I have a suggestion - you might wanna consider.
Make the links look like this:  https://app.asana.com/0/135618695592030/170130300261913
Pattern: https://app.asana.com/0/project-id/task-id
This opens project view of task, the one you used opens "My Tasks".
Which one will be better? Your call.

ShunSakurai commented:
Thank you very much for your feedback. Please find my answers below. 

**Run a lint and move query parameters to AsanaController.js**
Well noted. I will just keep the lint part of this commit. 

**Make the project not a required field** and
**Default me as the assignee when creating a task without assignee, project, or tag**
Now I understood your design philosophy.
Actually I use Asana more personally. Most often I use the extension for bookmarking Web pages, and I just complete or delete the task after reading the article. Other use case is to create a task in My Tasks list and directly drag it to other task as its subtask. I feel it slow when I have to select a project all the time.
However, I do understand that some Asana beginners don't understand the principle "Tasks are always part of project" and it leads to difficulty. Your approach to fix this is valuable.
How about making it a users' choice whether or not to make project as the required field? I can place the on/off button in the Utility page you can find in my [utility branch](https://github.com/ShunSakurai/asana-chrome-plugin/tree/utility).
This is not a strong objection. It's your repository, I will respect your decisions.

**Open the page by clicking the successful task creation status**
I just used /0/0/task-id format because I didn't make project compulsory in my version, but it's no problem using the https://app.asana.com/0/project-id/task-id format.
If the project becomes the optional field, it would be something like `$scope.taskCreationStatus.parent-id = project-id || user-id || tag-id || 0`.
BTW, https://app.asana.com/0/0/task-id doesn't always open in My Tasks list. I tested with a few tasks, and it seems Asana first tries to display the task in project page, then in tag page, and if both are not added to the task, lastly in the user page.

**Add link to the Chrome Web Store page**
I forgot to mention this in my previous comment (it's edited now). Is it OK to add the link in README?

ShunSakurai commented:
Thank you very much for merging! I will share some comments I was drafting.

As requested, I have put back the query parameters to AsanaGatewayService.js.
Also, I changed the URL format to https://app.asana.com/0/container-id/task-id. More specifically, I wrote:
`container_id: (response.projects[0])? response.projects[0].id: (response.tags[0])? response.tags[0].id: (response.assignee)? response.assignee.id: 0`
This is just explicitly writing down the order Asana seems to be using inexplicitly.

As for making the projects not required, I decided to remain a little more pushy.
In my opinion, assigning a task to someone is as good as adding it to a project, to keep its track. Tags are weak in this field, so I removed tags from the "Default to me" commit.
How to use Asana varies depends on each team or person. Some use projects intensively, and some use assignment heavily. One example of the latter is the simplest email address, x@mail.asana.com. After sending to this email, we take actions such as adding to a project, dragging into some task's subtasks, or just complete. Quickly adding a "task assigned to me" itself turns a reminder and it gives us time margin to act on the task later.
I wanted to have this tool beneficial to both types of users.

One issue with using Create Task function is, [we cannot select assignee in Personal Projects](http://stackoverflow.com/questions/27613418/how-to-get-list-of-users-in-personal-projects-workspace-using-asana-api). Defaulting the user provides the users a way to assign tasks to themselves. Although [Personal Projects is discontinued](https://blog.asana.com/2012/06/introducing-personal-projects-easy-task-sharing-for-friends-and-family/), I assume there are still many light users who uses My Tasks page in Personal Projects only.

Thanks again for understanding it. I see you are creating some settings features. I think "make the projects required" option can be included on that page, giving the users a chance to personalize this tool.
